### ISSUE TITLE
Generated table names truncated when they are too long   (originally: Many-to-Many association migration crash with 'alter' - trailing underscore in table name)

### ISSUE THREAD

**Issue by sgress454** [balderdashy/sails#4011]
_From @tibotiber on April 30, 2015 2:48_

The error obtained is:

```
Waterline encountered a fatal error when trying to perform the `alter` auto-migration strategy.
In a couple of seconds, the data (cached in memory) will be logged to stdout.
(a failsafe put in place to preserve development data)

In the mean time, here's the error:

Error (E_UNKNOWN) :: Encountered an unexpected error
error: relation "neighborhood_activatedservices__service_activatedneighborhoods_" does not exist
    at Connection.parseE (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:534:11)
    at Connection.parseMessage (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:361:17)
    at Socket.<anonymous> (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:105:22)
    at Socket.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:764:14)
    at Socket.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:426:10)
    at emitReadable (_stream_readable.js:422:5)
    at readableAddChunk (_stream_readable.js:165:9)
    at Socket.Readable.push (_stream_readable.js:127:10)
    at TCP.onread (net.js:528:21)
```

with the models as follow:

```
// Service.js
activatedNeighborhoods: {
    collection: 'neighborhood',
    via: 'activatedServices'
},

// Neighborhood.js
activatedServices: {
    collection: 'service',
    via: 'activatedNeighborhoods'
},
```

(edit) It only crashes when there is a value in the join table

_Copied from original issue: balderdashy/sails-postgresql#157_

**Comment by sgress454** [comment]
_From @tibotiber on April 30, 2015 3:53_

OK after some more testing, removing the camelCase hasn't solved the bug. Any idea? Why is this trailing underscore doing there in the table name? `neighborhood_activatedservices__service_activatedneighborhoods_`

**Comment by sgress454** [comment]
_From @tibotiber on April 30, 2015 4:0_

Extra info in case it's useful:

```
================================
Data backup:
================================

[ { id: 2,
    neighborhood_activatedServices: 1,
    service_activatedNeighborhoods: 1 } ]
error: A hook (`orm`) failed to load!
error: Error (E_UNKNOWN) :: Encountered an unexpected error
error: relation "neighborhood_activatedservices__service_activatedneighborhoods_" does not exist
    at Connection.parseE (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:534:11)
    at Connection.parseMessage (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:361:17)
    at Socket.<anonymous> (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:105:22)
    at Socket.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:764:14)
    at Socket.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:426:10)
    at emitReadable (_stream_readable.js:422:5)
    at readableAddChunk (_stream_readable.js:165:9)
    at Socket.Readable.push (_stream_readable.js:127:10)
    at TCP.onread (net.js:528:21)

Details:  error: relation "neighborhood_activatedservices__service_activatedneighborhoods_" does not exist

error: Error (E_UNKNOWN) :: Encountered an unexpected error
error: relation "neighborhood_activatedservices__service_activatedneighborhoods_" does not exist
    at Connection.parseE (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:534:11)
    at Connection.parseMessage (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:361:17)
    at Socket.<anonymous> (/home/tibo/ubismart/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:105:22)
    at Socket.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:764:14)
    at Socket.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:426:10)
    at emitReadable (_stream_readable.js:422:5)
    at readableAddChunk (_stream_readable.js:165:9)
    at Socket.Readable.push (_stream_readable.js:127:10)
    at TCP.onread (net.js:528:21)

Details:  error: relation "neighborhood_activatedservices__service_activatedneighborhoods_" does not exist
```

**Comment by sgress454** [comment]
_From @sailsbot on September 23, 2015 5:56_

Thanks for posting, @tibotiber.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:
- review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
- create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sgress454** [comment]
_From @tibotiber on September 24, 2015 4:28_

FYI, switched to `sails-migrations` now to avoid such issue.

**Comment by sgress454** [comment]
_From @sailsbot on October 25, 2015 5:8_

Thanks for posting, @tibotiber.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:
- review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
- create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sgress454** [comment]
_From @iit-federico-bozzini on November 26, 2015 17:38_

I encountered the same problem.

Is this going to be fixed?

**Comment by sgress454** [comment]
_From @particlebanana on January 25, 2016 23:27_

This has to do with the way the generated table names are created. It's a combination of the tablename and attribute name. In this case the names are longer than the allowable table name.

For SQL adapters manual migrations are always recommended. It can be hard with join tables though.

This has been discussed a few times #139, https://github.com/balderdashy/waterline/issues/725, and https://github.com/balderdashy/waterline-schema/issues/20 it looks like without a solution.

We need to come up with an actual plan for this.

**Comment by sgress454** [comment]
The workaround for this for now is to declare the join table name yourself by using a `through` model:

```
// Service.js
attributes: {
  activatedNeighborhoods: {
      collection: 'neighborhood',
      via: 'service',
      through: 'NeighborhoodServices'
  }
}

// Neighborhood.js
attributes: {
  activatedServices: {
      collection: 'service',
      via: 'neighborhood',
      through: 'NeighborhoodServices'
  }
}

// NeighborhoodServices.js
tableName: 'neighborhood_services',
attributes: {
  neighborhood: {
      model: 'neighborhood'
  }, 
  service: {
      model: 'service'
  }
}
```
